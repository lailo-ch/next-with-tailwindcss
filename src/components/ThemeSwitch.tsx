import * as React from 'react'
import Switch from 'react-switch'
import { FiMoon, FiSun } from 'react-icons/fi'

const DARK_MODE_CLASS_NAME = 'dark-mode'
const DARK_MODE_TRANSITION_CLASS_NAME = 'dark-mode-transition'
const DARK_MODE_TRANSITION_DURATION = 500

const ThemeSwitch: React.FC = () => {
  const [hasActiveDarkMode, activateDarkMode] = React.useState(false)

  const toggleDarkMode = () => {
    document.documentElement.classList.add(DARK_MODE_TRANSITION_CLASS_NAME)
    if (hasActiveDarkMode) {
      document.documentElement.classList.remove(DARK_MODE_CLASS_NAME)
    } else {
      document.documentElement.classList.add(DARK_MODE_CLASS_NAME)
    }
    activateDarkMode(!hasActiveDarkMode)
    setTimeout(
      () =>
        document.documentElement.classList.remove(
          DARK_MODE_TRANSITION_CLASS_NAME
        ),
      DARK_MODE_TRANSITION_DURATION
    )
  }

  React.useEffect(() => {
    if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
      toggleDarkMode()
    }
  }, [])

  return (
    <Switch
      onChange={toggleDarkMode}
      checked={hasActiveDarkMode}
      checkedIcon={<FiSun className="inline-block mx-2" />}
      uncheckedIcon={<FiMoon className="inline-block mx-2" />}
      onColor="#f7fafc"
      offColor="#1a202c"
      onHandleColor="#1a202c"
      offHandleColor="#ffffff"
      className="text-white dark:text-black"
    />
  )
}

export default ThemeSwitch
