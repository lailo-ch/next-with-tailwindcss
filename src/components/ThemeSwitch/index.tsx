import * as React from 'react'
import Switch from 'react-switch'

const DARK_MODE_KEY = 'dark-mode'

const ThemeSwitch: React.FunctionComponent = () => {
  const [hasActiveDarkMode, activateDarkMode] = React.useState<boolean>(false)

  const toggleDarkMode = () => {
    if (hasActiveDarkMode) {
      document.body.classList.remove(DARK_MODE_KEY)
    } else {
      document.body.classList.add(DARK_MODE_KEY)
    }
    activateDarkMode(!hasActiveDarkMode)
  }

  React.useEffect(() => {
    if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
      toggleDarkMode()
    }
  }, [])

  return (
    <Switch
      onChange={toggleDarkMode}
      checked={hasActiveDarkMode}
      checkedIcon={null}
      uncheckedIcon={null}
      onColor="#f7fafc"
      offColor="#1a202c"
      onHandleColor="#1a202c"
      offHandleColor="#ffffff"
    />
  )
}

export default ThemeSwitch
