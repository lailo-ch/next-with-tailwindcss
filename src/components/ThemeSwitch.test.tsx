import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import ThemeSwitch from './ThemeSwitch'

describe('ThemeSwitch', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<ThemeSwitch />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
