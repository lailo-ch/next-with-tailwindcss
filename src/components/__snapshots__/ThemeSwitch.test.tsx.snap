// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitch matches snapshot 1`] = `
<ReactSwitch
  activeBoxShadow="0 0 2px 3px #3bf"
  boxShadow={null}
  checked={false}
  checkedIcon={
    <FiSun
      className="inline-block mx-2"
    />
  }
  className="text-white dark:text-black"
  disabled={false}
  height={28}
  offColor="#1a202c"
  offHandleColor="#ffffff"
  onChange={[Function]}
  onColor="#f7fafc"
  onHandleColor="#1a202c"
  uncheckedIcon={
    <FiMoon
      className="inline-block mx-2"
    />
  }
  width={56}
/>
`;
